### PR TITLE
feat(datepicker): add datepicker field

### PR DESCRIPTION
**Pull request by edetec**

### CODE REVIEW

File: demo/src/index.html [hunk 1]
```diff
@@ -7,31 +7,30 @@
     @esss/ng-xform
   </title>
   <base href=".">
-  <meta name="description" content="Angular library built with ❤ using ngx-library yeoman generator." />
```
Comment (lvoliveira) [inline demo/src/index.html:10]: This is no longer built with heart? :'(

File: src/ng-xform/fields/datepicker-field.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import { DynamicField } from './dynamic-field';
+
+export class DatepickerField extends DynamicField {
```
Comment (igortg) [inline src/ng-xform/fields/datepicker-field.ts]: Name should be `DateField`. Let's keep ng-xform in a higher level of abstraction. User just need a field for a "date". Doesn't mind if is a "picker", a "swiper" or other kind of interaction.

File: src/ng-xform/ng-xform.component.spec.ts [hunk 3]
```diff
@@ -178,12 +181,32 @@ describe('DynamicFormComponent', () => {
...
   });
 
+  it('should patch value', () => {
+    component.createForm();
+    component.reset();
+    expect(component.errorCode).toBeUndefined();
+    expect(component.form.value.nested2).toBeTruthy();
+  });
+
+  it('should change nested attr value', () => {
```
Comment (igortg) [inline src/ng-xform/ng-xform.component.spec.ts:191]: nice

File: src/ng-xform/datepicker-field/datepicker-field.component.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
+import { BehaviorSubject } from 'rxjs/BehaviorSubject';
+import { ControlValueAccessor, FormControl, NG_VALUE_ACCESSOR } from '@angular/forms';
+import { Component, AfterContentInit, ViewChild, ElementRef, Renderer, AfterViewInit } from '@angular/core';
+import { BsDatepickerConfig, BsLocaleService } from 'ngx-bootstrap/datepicker';
+
+import { BaseDynamicFieldComponent } from '../field-components/base-dynamic-field.component';
+import { DatepickerField } from '../fields';
+
+/**
+ * Component to generate a bootstrap form field of general type
```
Comment (igortg) [inline src/ng-xform/datepicker-field/datepicker-field.component.ts]: wrong doc

File: src/ng-xform/datepicker-field/datepicker-field.component.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+@Component({
+  selector: 'ng-xform-datepicker-field',
+  templateUrl: './datepicker-field.component.html',
+  providers: [{
+    provide: NG_VALUE_ACCESSOR,
+    useExisting: DatepickerFieldComponent,
+    multi: true
+  }],
+})
+export class DatepickerFieldComponent extends BaseDynamicFieldComponent<DatepickerField> implements AfterViewInit, AfterContentInit,
```
Comment (igortg) [inline src/ng-xform/datepicker-field/datepicker-field.component.ts]: Should be `DateFieldComponent`

File: src/ng-xform/datepicker-field/datepicker-field.component.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  ControlValueAccessor {
+  config: Partial<BsDatepickerConfig>;
+  componentControl = new FormControl();
+  private input: HTMLInputElement;
+
+  _onChange = (value: any) => { };
+  _onTouched = () => { };
+
+  constructor(private _localeService: BsLocaleService, private elementRef: ElementRef, private renderer: Renderer) {
```
Comment (igortg) [inline src/ng-xform/datepicker-field/datepicker-field.component.ts]: Do we need the "_" prefix?

File: src/ng-xform/datepicker-field/datepicker-field.component.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    multi: true
+  }],
+})
+export class DatepickerFieldComponent extends BaseDynamicFieldComponent<DatepickerField> implements AfterViewInit, AfterContentInit,
+  ControlValueAccessor {
+  config: Partial<BsDatepickerConfig>;
+  componentControl = new FormControl();
+  private input: HTMLInputElement;
+
+  _onChange = (value: any) => { };
```
Comment (lvoliveira) [inline src/ng-xform/datepicker-field/datepicker-field.component.ts]: EventEmiter would not work?

File: src/ng-xform/datepicker-field/datepicker-field.component.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  ngAfterViewInit() {
+    this.input = this.elementRef.nativeElement.querySelector('input');
+    if (this.input) {
+      this.input.onblur = this._onTouched
+    }
+  }
+
+  writeValue(obj: any): void {
```
Comment (igortg) [inline src/ng-xform/datepicker-field/datepicker-field.component.ts]: Weird? The component doesn't always return `Date`?

File: src/ng-xform/datepicker-field/datepicker-field.component.ts [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    });
+  }
+
+  ngAfterContentInit() {
+    this.componentControl.setValidators(this.field.validators);
+    this.config = Object.assign({}, {
+      containerClass: `theme-${this.field.theme}`,
+    });
+    if (this.field.locale) {
+      this._localeService.use(this.field.locale)
```
Comment (igortg) [inline src/ng-xform/datepicker-field/datepicker-field.component.ts]: I couldn't find where the component is using `_localeService`

File: src/ng-xform/datepicker-field/datepicker-field.component.html [hunk 10]
```diff
@@ -0,0 +1,15 @@
+<div *ngIf="!editing" id="{{field.key}}-label">
+  <strong>{{ field.label }}</strong>
+  <p class="text-muted">{{ form?.controls[field.key]?.value | date:'mediumDate':field.locale }}</p>
+</div>
+
+<div *ngIf="editing" id="{{elementId}}-div" class="form-group" [formGroup]="form" [ngClass]="displayFieldCss()">
+  <label for="{{elementId}}">{{field.label}}</label>
+  <input type="text" class="form-control" [formControl]="componentControl" id="{{elementId}}" bsDatepicker
```
Comment (igortg) [inline src/ng-xform/datepicker-field/datepicker-field.component.html]: `bsDatepicker` directive should be right after the `<input`, sou we could easily identify that is a Datepicker and not a text input

File: src/ng-xform/datepicker-field/datepicker-field.component.ts [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+      this.componentControl.setValue(new Date(obj));
+    }
+  }
+  registerOnChange(fn: any): void {
+    this._onChange = fn;
+  }
+  registerOnTouched(fn: any): void {
+    this._onTouched = fn;
+  }
+  setDisabledState?(isDisabled: boolean): void {
```
Comment (igortg) [inline src/ng-xform/datepicker-field/datepicker-field.component.ts]: Why this is necessary?

File: src/ng-xform/datepicker-field/datepicker-field.component.ts (hunk 6 above)
Comment (igortg) [inline src/ng-xform/datepicker-field/datepicker-field.component.ts]: Explain (in the docs):
- Why are you injecting BsLocaleService?
- Why are you injecting elementRef?

File: demo/src/app/load-locales.ts [hunk 12]
```diff
@@ -0,0 +1,16 @@
+import { defineLocale, LocaleData } from 'ngx-bootstrap/chronos';
```
Comment (igortg) [inline demo/src/app/load-locales.ts]: Explain (with docs) why we need that

**Review by igortg — CHANGES_REQUESTED** [review]

File: src/ng-xform/datepicker-field/datepicker-field.component.ts (hunk 7 above)
Comment (edetec) [reply src/ng-xform/datepicker-field/datepicker-field.component.ts]: yep, but only one handler is necessary

File: src/ng-xform/fields/datepicker-field.ts (hunk 2 above)
Comment (edetec) [reply src/ng-xform/fields/datepicker-field.ts]: OK

File: src/ng-xform/datepicker-field/datepicker-field.component.ts (hunk 4 above)
Comment (edetec) [reply src/ng-xform/datepicker-field/datepicker-field.component.ts]: OK

File: src/ng-xform/datepicker-field/datepicker-field.component.ts (hunk 6 above)
Comment (edetec) [reply src/ng-xform/datepicker-field/datepicker-field.component.ts]: OK

File: src/ng-xform/datepicker-field/datepicker-field.component.ts (hunk 9 above)
Comment (edetec) [reply src/ng-xform/datepicker-field/datepicker-field.component.ts]: this is the service used by third-party DatePickerInput Directive

File: src/ng-xform/datepicker-field/datepicker-field.component.html (hunk 10 above)
Comment (edetec) [reply src/ng-xform/datepicker-field/datepicker-field.component.html]: OK

File: src/ng-xform/datepicker-field/datepicker-field.component.ts (hunk 11 above)
Comment (edetec) [reply src/ng-xform/datepicker-field/datepicker-field.component.ts]: To inform to the form when on blur occur on the component. The angular Form uses it to update your state.

File: demo/src/app/load-locales.ts (hunk 12 above)
Comment (igortg) [reply demo/src/app/load-locales.ts]: Do the locale review for this component, as discussed in person.

File: src/ng-xform/datepicker-field/datepicker-field.component.ts (hunk 6 above)
Comment (edetec) [reply src/ng-xform/datepicker-field/datepicker-field.component.ts]: OK

File: demo/src/app/load-locales.ts (hunk 12 above)
Comment (edetec) [reply demo/src/app/load-locales.ts]: this is explained in README.md on section Locales

**Pull request merged**